File: Medium/2130. Maximum Twin Sum of a Linked List/Solution.js
var reverseList = function (head) {
    if (head == null || head.next == null) return head;
    let cur = head.next;
    head.next = null;
    while (cur != null) {
        let go = cur.next;
        cur.next = head;
        head = cur;
        cur = go;
    }
    return head;
};

var pairSum = function(head) {
    let fast = head, slow = head;
    while (fast) {
        fast = fast.next.next;
        slow = slow.next;
    }

    let startReversed = reverseList(slow);
    let ans = 0;
    while (startReversed) {
        ans = Math.max(ans, head.val + startReversed.val);
        head = head.next;
        startReversed = startReversed.next;
    }
    return ans;
};

File: Medium/2130. Maximum Twin Sum of a Linked List/Solution.cpp
class Solution {
public:
    ListNode* reverseList(ListNode* head) {
        if(!head or !head->next) return head;
        auto *cur = head->next;
        head->next = nullptr;
        while(cur) {
            auto *go = cur->next;
            cur->next = head;
            head = cur;
            cur = go;
        }
        return head;
    }

    int pairSum(ListNode* head) {
        auto *fast = head, *slow = head;
        while(fast) {
            fast = fast->next->next;
            slow = slow->next;
        }

        auto *startReversed = reverseList(slow);
        int ans = 0;
        while(startReversed) {
            ans = max(ans, head->val + startReversed->val);
            head = head->next;
            startReversed = startReversed->next;
        }
        return ans;
    }
};